File: 401. Binary Watch/readBinaryWatch.js
/**
 * @param {number} turnedOn
 * @return {string[]}
 */
var readBinaryWatch = function (turnedOn) {
    //hours     => 0->11 => represent: 8  4  2  1
    //minutes   => 0->59 => represent: 32 16 8  4  2  1

    const hrs = [1, 2, 4, 8] //len = 4  //max value can be form = 11        //TOTAL = 15
    const mins = [1, 2, 4, 8, 16, 32] //len = 6 //max value can be form = 59 //TOTAL = 63
    let ans = new Set()

    function createTime(total, idx, hr, min) {
        if (hr > 11 || min > 59) return
        if (total === 0) {
            let form = min < 10 ? "0" + min : min
            ans.add(`${hr}:${form}`)
            return;
        }
        for (let i = idx; i < 10; i++) {
            let currHr = hr
            let currMin = min
            //Max hrs Length = 4
            if (i < 4) {
                currHr += hrs[i]
            }
            else {
                currMin += mins[i - 4]
            }
            createTime(total - 1, i + 1, currHr, currMin)
        }

    }

    createTime(turnedOn, 0, 0, 0)

    return [...ans]
};

File: 401. Binary Watch/readBinaryWatch.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    vector<int> mins = {1, 2, 4, 8, 16, 32};
    vector<int> hrs = {1, 2, 4, 8};
    void createTime(vector<string>& ans, int total, int idx, int hr, int min){
        if(hr > 11 || min > 59)return;
        if(total == 0){
            string form = min < 10 ? "0"+to_string(min) : to_string(min);
            ans.push_back(to_string(hr)+":"+form);
        }

        for(int i=idx; i<10; i++){
            int currHr = hr;
            int currMin = min;

            if(i < 4){
                currHr += hrs[i];
            }
            else{
                currMin += mins[i-4];
            }
            createTime(ans, total-1, i+1, currHr, currMin);
        }
    }
public:
    vector<string> readBinaryWatch(int turnedOn) {
        vector<string> ans;

        createTime(ans, turnedOn, 0, 0, 0);

        return ans;
    }
};